library/artifactory_security.py

```diff
@@ -0,0 +1,570 @@
+    state: absent
+'''
+
+RETURN = '''
+original_message:
+    description:
+        - A brief sentence describing what action the module was attempting
+          to take against which user/group/permission and what artifactory url.
+    returned: success
+    type: str
+message:
+    description: The result of the attempted action.
+    returned: success
+    type: str
+config:
+    description:
+        - The configuration of a successfully created user/group/permission,
+          an updated user/group/permission (whether or not changed=True), or
+          the config of a user/group/permission that was successfully deleted.
+    returned: success
+    type: dict
+'''
+
+
+import ast
+import json
+
+import ansible.module_utils.artifactory as art_base
+import ansible.module_utils.six.moves.urllib.error as urllib_error
+
+from ansible.module_utils.basic import AnsibleModule
+
+
+USER_CONFIG_MAP = {
+    "email":
+        {"always_required": True},
+    "password":
+        {"always_required": True}}
+
+PERMISSION_CONFIG_MAP = {
+    "repositories":
+        {"always_required": True}}
+
+URI_CONFIG_MAP = {"api/security/users": USER_CONFIG_MAP}
+URI_CONFIG_MAP["api/security/permissions"] = PERMISSION_CONFIG_MAP
+# There are no required values for groups, but if they are not defined
+# a validation error will be thrown.
+URI_CONFIG_MAP["api/security/groups"] = True
+
+
+class ArtifactorySecurity(art_base.ArtifactoryBase):
+    def __init__(self, artifactory_url, name=None,
+                 sec_config=None, username=None, password=None,
+                 auth_token=None, validate_certs=False, client_cert=None,
+                 client_key=None, force_basic_auth=False, config_map=None):
+        super(ArtifactorySecurity, self).__init__(
+            username=username,
+            password=password,
+            auth_token=auth_token,
+            validate_certs=validate_certs,
+            client_cert=client_cert,
+            client_key=client_key,
+            force_basic_auth=force_basic_auth,
+            config_map=config_map)
+        self.artifactory_url = artifactory_url
+        self.name = name
+        self.sec_config = sec_config
+
+        if self.name:
+            self.working_url = '%s/%s' % (self.artifactory_url, self.name)
+        else:
+            self.working_url = self.artifactory_url
+
+    def get_targets(self):
+        return self.query_artifactory(self.artifactory_url, 'GET')
+
+    def get_target_config(self):
+        return self.query_artifactory(self.working_url, 'GET')
+
+    def delete_target(self):
+        return self.query_artifactory(self.working_url, 'DELETE')
+
+    def create_target(self):
+        method = 'PUT'
+        serial_config_data = self.get_valid_conf(method)
+        create_target_url = self.working_url
+        return self.query_artifactory(create_target_url, method,
+                                      data=serial_config_data)
+
+    def update_target_config(self):
+        method = 'POST'
```

Comment:
for permissions targets their is no update. you have to use [Create or Replace Permissions Target](https://www.jfrog.com/confluence/display/RTF/Artifactory+REST+API#ArtifactoryRESTAPI-CreateorReplacePermissionTarget) which means a PUT instead of POST.

Reply:
I was hoping they would have an informative error message and I'd just pass that back up the chain, but apparently that's too difficult. It just throws a 400 with "Bad Request" as the message. Disappointing.